remove path from failpoint

lavacat opened the pull request:
See https://github.com/etcd-io/etcd/issues/14727

serathius commented:
This looks good for me. cc @ahrtr who had different opinion about the topic. 
Can you fix the DCO check? Just run `git checkout dd4869804163318da0745652c38501072f3c824d && git commit --amend --signoff && git push HEAD:remove-fp-path -f`

ahrtr commented:
> cc @ahrtr who had different opinion about the topic.

Just as I mentioned in https://github.com/etcd-io/etcd/issues/14727#issuecomment-1329654978, it isn't a big deal. Since no feedback from other maintainers. I don't want to be blocked here. I agree to remove the pacakge path, although personally I tend to keep it.

ahrtr commented on runtime/runtime.go:
```diff
@@ -147,11 +147,17 @@ func list() []string {
 
-func register(failpath string, fp *Failpoint) {
+func register(name string, fp *Failpoint) {
 	failpointsMu.Lock()
```
Note that we remove the lock for each failpoint. We need to acquire the global lock for each operation.

ahrtr commented on runtime/runtime.go:
```diff
@@ -147,11 +147,17 @@ func list() []string {
+			name)
+	}
+	failpoints[name] = fp
 	failpointsMu.Unlock()
```
```suggestion
```

ahrtr commented:
We don't have e2e test for now, please make sure you manually verify the PR using [examples](https://github.com/etcd-io/gofail/tree/master/examples)

serathius commented:
@lavacat Can you apply the suggestions by @ahrtr ?

lavacat commented:
- added unit tests 
- renamed `failpath` to `name` everywhere in runtime.go

@ahrtr FYI with your suggestion there was a deadlock, because register was calling `Enable` that acquires the same lock. Changed it to call `enable`. I think that was a typo in original code.

lavacat commented:
- changed NewFailpoint/register logic to return existing failpoint if name is already registered
- updated readme
- minor unit tests improvements
@serathius @ahrtr I think this is ready to merge. Please take a look.

ahrtr commented on runtime/runtime.go:
```diff
@@ -147,11 +147,22 @@ func list() []string {
-		Enable(failpath, t)
+	defer failpointsMu.Unlock()
+	if f, ok := failpoints[name]; ok {
+		fmt.Fprintf(
+			os.Stderr,
+			"warning: failpoint name %s is already registered. You are using same failpoint name in multiple packages.\n",
+			name)
+		// return existing failpoint, don't re-enable
+		return f
```
We shouldn't allow duplicated failpoint, so we might need to panic here? WDYT? cc @serathius

serathius replied:
Yep, we definitely need to panic here.

ahrtr commented:
Overall looks good to me with two comments:
1. Please fix the test failure;
2. Please consider using `github.com/stretchr/testify` in the unit test so simplify the implementation.

lavacat commented:
> Overall looks good to me with two comments:
> 
> 1. Please fix the test failure;
> 2. Please consider using `github.com/stretchr/testify` in the unit test so simplify the implementation.

@ahrtr thanks, addressed both

ahrtr commented on code/binding_test.go:
```diff
@@ -0,0 +1,39 @@
+package code
+
+import (
+	"bytes"
+	"github.com/stretchr/testify/assert"
+	"testing"
```
```suggestion
	"bytes"
	"testing"
	
	"github.com/stretchr/testify/assert"
```

ahrtr commented on runtime/failpoint_test.go:
```diff
@@ -0,0 +1,53 @@
+package runtime
+
+import (
+	"github.com/stretchr/testify/assert"
+	"testing"
```
```suggestion
	"testing"
	
	"github.com/stretchr/testify/assert"
```

serathius commented on runtime/runtime.go:
```diff
@@ -147,11 +147,19 @@ func list() []string {
+	defer failpointsMu.Unlock()
+	if _, ok := failpoints[name]; ok {
+		panic(fmt.Sprintf(
+			"failpoint name %s is already registered. You are using same failpoint name in multiple packages.\n",
```
```suggestion
			"failpoint name %s is already registered.",
```
Could be same package.

ahrtr reviewed (approved):
LGTM

Thank you @lavacat
